### ISSUE TITLE
create a dependency graph to speed up dependency evaluation

### ISSUE THREAD

**Issue by yoon** [NUARIG/surveyor#419]
like in NUBIC/nu_surveyor#7

**Comment by yoon** [comment]
Although there is a proposal to achieve speed by moving all dependency evaluation to the client, that would involve making the client aware of responses in all sections. An alternative has been discussed - calculate a dependency graph when the survey is saved and store it in the database. The latter would be easier to implement quickly, and would preserve sqlite compatibility.

**Comment by rsutphin** [comment]
I agree that there is much room for improving dependency evaluation on the server side before re-implementing it all on the client side. Additionally, there are non-client applications for dependency evaluation. For example, you might have an application that wants to determine if all "required" questions have been answered, except for those which were not presented due to a dependency hiding them. For this to work efficiently, the ruby-side implementation of dependencies will need to be improved anyway.

**Comment by rsutphin** [comment]
If dependency evaluation is updated with this ticket, it might be good to take the use case under #366 into account.

**Comment by bigtunacan** [comment]
I see you are looking at a server side solution; just wanted to post a heads up that we have moved dependency evaluation to the client side on our fork of the project.  If a client side solution is still being considered we could look at doing a pull request.

**Comment by flockom** [comment]
@bigtunacan are you willing to share your fork? I did not find it with a quick search.

**Comment by bigtunacan** [comment]
@flockom Yes; it's under another account.  The client side dependency evaluation is currently residing within a project we have built on top of surveyor, rather than within the gem itself.  I'll make some updates on my end to push those changes down into the gem, and then I will make the fork available.  It will likely be sometime next week before it's ready.
